File: 5/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = Infinity;

const rawSeeds = input
  .split("\n")[0]
  .split(": ")[1]
  .split(" ")
  .map((d) => parseInt(d));
// const seeds = [79];

let converters = input
  .split("\n\n")
  .slice(1)
  .map((d) =>
    d
      .split("\n")
      .slice(1)
      .map((d) => d.split(" ").map((d) => parseInt(d)))
  );

while (rawSeeds.length) {
  const from = rawSeeds.shift()!;
  const l = rawSeeds.shift()!;
  for (let i = 0; i < l; i++) {
    let seed = from + i;
    for (const map of converters) {
      for (const [to, from, l] of map) {
        if (from <= seed && from + l > seed) {
          seed = to + (seed - from);
          break;
        }
      }
    }
    // console.log(seed);
    if (seed < res) res = seed;
  }
}

// converters = [converters.shift()!];

// console.log(seeds);

console.log();
console.log(res);

File: 5/b.rs
use rayon::prelude::*;
use std::fs;

#[derive(Debug)]
struct Translation {
  to: u64,
  from: u64,
  length: u64,
}

impl Translation {
  fn new(str: &str) -> Self {
    let mut parts = str.split(" ");
    let to = parts.next().unwrap().parse::<u64>().unwrap();
    let from = parts.next().unwrap().parse::<u64>().unwrap();
    let length = parts.next().unwrap().parse::<u64>().unwrap();
    Self { to, from, length }
  }
}

fn main() {
  let file = fs::read_to_string("input.txt").unwrap();

  let mut blocks = file.split("\n\n");
  let raw_seeds = blocks
    .next()
    .unwrap()
    .strip_prefix("seeds: ")
    .unwrap()
    .split(" ")
    .map(|x| x.parse::<u64>().unwrap())
    .collect::<Vec<u64>>();

  let steps = blocks
    .map(|block| {
      let mut translations = block.lines();
      translations.next();
      translations
        .map(|x| Translation::new(x))
        .collect::<Vec<Translation>>()
    })
    .collect::<Vec<Vec<Translation>>>();

  let res = raw_seeds
    .par_chunks(2)
    .flat_map(|x| {
      (x[0]..(x[0] + x[1])).into_par_iter().map(|mut num| {
        for step in &steps {
          for translation in step {
            if num >= translation.from && num < translation.from + translation.length {
              num = translation.to + (num - translation.from);
              break;
            }
          }
        }
        num
      })
    })
    .min()
    .unwrap();

  println!("{:?}", res);
}
